File: swara-music/src/main/java/com/swara/music/readers/MidiReader.java
package com.swara.music.readers;

import java.io.InputStream;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import javax.sound.midi.MetaMessage;
import javax.sound.midi.MidiEvent;
import javax.sound.midi.MidiSystem;
import javax.sound.midi.Sequence;
import javax.sound.midi.ShortMessage;

import com.google.common.base.Preconditions;
import com.swara.music.MusicReader;
import com.swara.music.elements.Chord;
import com.swara.music.elements.Fragment;
import com.swara.music.elements.Key;
import com.swara.music.elements.Note;
import com.swara.music.elements.Phrase;
import com.swara.music.elements.Song;
import com.swara.music.elements.Tempo;
import com.swara.music.elements.Voice;

import org.apache.commons.math3.fraction.Fraction;

/**
 * Reads a {@link Song} from a MIDI {@link javax.sound.midi.Sequence}.
 */
public class MidiReader implements MusicReader {

    private static final Function<Integer, Predicate<MidiEvent>> META_EVENT = type ->
        (evt -> evt.getMessage() instanceof MetaMessage && ((MetaMessage) evt.getMessage()).getType() == type);

    private static final Function<Integer, Predicate<MidiEvent>> CHANNEL_EVENT = command ->
        (evt -> evt.getMessage() instanceof ShortMessage && ((ShortMessage) evt.getMessage()).getCommand() == command);

    private static final Predicate<MidiEvent> MIDI_KEY  = META_EVENT.apply(0x59);
    private static final Predicate<MidiEvent> MIDI_BPM  = META_EVENT.apply(0x51);
    private static final Predicate<MidiEvent> MIDI_TIME = META_EVENT.apply(0x58);
    private static final Predicate<MidiEvent> MIDI_PROG = CHANNEL_EVENT.apply(0xC0);
    private static final Predicate<MidiEvent> MIDI_NOTE = CHANNEL_EVENT.apply(0x90).or(CHANNEL_EVENT.apply(0x80));

    @Override
    public Song read(InputStream in) throws Exception {
        // Read a MIDI sequence from the input stream and flatten all tracks.
        final Sequence sequence = MidiSystem.getSequence(in);
        Preconditions.checkArgument(sequence.getDivisionType() == Sequence.PPQ);
        final List<MidiEvent> events = Arrays.stream(sequence.getTracks())
            .flatMap(track -> IntStream.range(0, track.size()).mapToObj(track::get))
            .sorted((a, b) -> Long.compare(a.getTick(), b.getTick()))
            .collect(Collectors.toList());

        // Split the events whenever a new fragment begins.
        final TreeMap<Long, List<MidiEvent>> headers = events.stream()
            .filter(MIDI_KEY.or(MIDI_BPM).or(MIDI_TIME).or(MIDI_PROG))
            .collect(Collectors.groupingBy(MidiEvent::getTick, TreeMap::new, Collectors.toList()));

        // Iteratively construct a song.
        final Song.Builder song = new Song.Builder();
        final Key.Builder key = new Key.Builder();
        final Tempo.Builder tempo = new Tempo.Builder();
        final Map<Integer, Integer> programs = new HashMap<>();

        headers.keySet().forEach(start -> {
            final Fragment.Builder fragment = new Fragment.Builder();
            final List<MidiEvent> header = headers.get(start);

            // Read key information.
            header.stream().filter(MIDI_KEY).findAny()
                .map(i -> ((MetaMessage) i.getMessage()).getData())
                .ifPresent(i -> {
                    key.withSignature(i[0]);
                    key.withType(i[1]);
                });

            // Read tempo information.
            header.stream().filter(MIDI_BPM).findAny()
                .map(i -> ((MetaMessage) i.getMessage()).getData())
                .map(i -> 60_000_000 / ((i[0] << 16) + (i[1] << 8) + i[2]))
                .ifPresent(tempo::withBpm);

            header.stream().filter(MIDI_TIME).findAny()
                .map(i -> ((MetaMessage) i.getMessage()).getData())
                .ifPresent(i -> tempo.withSignature(i[0], (int) Math.pow(2, i[1])));

            // Read program change information.
            header.stream().filter(MIDI_PROG)
                .map(i -> (ShortMessage) i.getMessage())
                .forEach(msg -> programs.put(msg.getChannel(), msg.getData1()));

            // Parse the fragment channel information.
            events.stream().filter(MIDI_NOTE.and(i -> headers.floorKey(i.getTick()).equals(start)))
                .collect(Collectors.groupingBy(i -> ((ShortMessage) i.getMessage()).getChannel()))
                .forEach((channel, list) -> {
                    final Voice.Builder voice = new Voice.Builder();
                    final Map<Integer, ShortMessage> notes = new HashMap<>();

                    // Group note on and note off events by tick.
                    final TreeMap<Long, List<MidiEvent>> ticks = list.stream().collect(
                        Collectors.groupingBy(MidiEvent::getTick, TreeMap::new, Collectors.toList())
                    );

                    ticks.forEach((tick, group) -> {
                        // Articulate all notes (if any) in the current chord.
                        final Long last = ticks.lowerKey(tick);
                        final long diff = tick - (last == null ? start : last);
                        final long pulses = 4 * sequence.getResolution();
                        final long length = ((diff + 60/2) / 60) * 60;

                        if (diff >= 60) {
                            voice.withChord(new Chord.Builder()
                                .withDuration(new Fraction((double) length / pulses))
                                .withNotes(notes.values().stream()
                                    .map(i -> new Note.Builder()
                                        .withPitch(i.getData1() % 12)
                                        .withOctave(i.getData1() / 12)
                                        .build())
                                    .collect(Collectors.toSet()))
                                .withVolume((int) notes.values().stream()
                                    .mapToInt(ShortMessage::getData2)
                                    .average()
                                    .orElse(0))
                                .build());
                        }

                        // Add all new events and remove all old events.
                        for (MidiEvent event : group) {
                            final ShortMessage msg = (ShortMessage) event.getMessage();
                            notes.merge(msg.getData1(), msg, (v1, v2) -> null);
                        }
                    });

                    fragment.withPhrase(channel, new Phrase.Builder()
                        .withVoice(voice.build())
                        .withProgram(programs.get(channel))
                        .build());
                });

            // Build the fragment and add it to the song.
            song.withFragment(fragment
                .withKey(key.build())
                .withTempo(tempo.build())
                .build());
        });

        // Assemble and return the parsed song.
        return song.build();
    }

}

File: swara-music/src/main/scala/com/swara/music/readers/MidiReader.scala
package com.swara.music
package readers

import java.io.InputStream
import javax.sound.midi._
import scala.util.Try
import scala.collection.mutable

/**
 * Reads a [[Song]] from a Midi [[javax.sound.midi.Sequence]]. Midi files provide a compact
 * alternative to space inefficient audio files. Midi files are composed of sequences of
 * time-stamped Midi messages, which each encode the various musical events that occur throughout a
 * song (e.g. key and tempo changes, note played, note released, etc.). The ability to read Midi
 * files enables the library to interface with conventional musical production and playback tools.
 * For more information on Midi, check out: http://www.somascape.org/midi/tech/mfile.html.
 */
object MidiReader extends MusicReader {

  override def read(in: InputStream): Try[Song] = Try {
    // Flatten all midi messages into a sorted map from midi tick to message.
    val sequence = MidiSystem.getSequence(in)
    require(sequence.getDivisionType == Sequence.PPQ, "Sequence must be divided in PPQ.")

//    var multiple = sequence.getResolution; while (multiple % 2 == 0) multiple /= 2
//    def round(x: Long, to: Double): Long = (math.round(x / to) * to).toLong

    val midi = mutable.TreeMap.empty[Long, Array[MidiMessage]]
    midi ++= sequence.getTracks
      .flatMap(t => (0 until t.size).map(t.get))
      .groupBy(_.getTick)
      .mapValues(_.map(_.getMessage))

    // Iteratively construct the various musical elements; defaults specified in midi format spec.
    val fragments = mutable.Buffer.empty[Fragment]
    var key = Key.CMajor
    var tempo = Tempo(Duration(4, 4), 120.0)
    val programs = mutable.Map.empty[Int, Int]
    val channels = mutable.Map.empty[Int, Phrase]

    while (midi.nonEmpty) {
      val (start, head) = midi.head

      // Update the fragment's key, time, tempo, and programs.
      head.collect {
        case m: MetaMessage if m.getType == 0x59 /* Key Change */ =>
          key = key.copy(signature = m.getData()(0), isMajor = m.getData()(1) == 1)
        case m: MetaMessage if m.getType == 0x58 /* Time Change */ =>
          tempo = tempo.copy(signature = Duration(m.getData()(0), Math.pow(2, m.getData()(1)).toInt))
        case m: MetaMessage if m.getType == 0x51 /* Tempo Change */ =>
          val mspqn = (m.getData()(0) << 16) + (m.getData()(1) << 8) + m.getData()(2)
          tempo = tempo.copy(bpm = 6E7 / mspqn)
        case m: ShortMessage if m.getCommand == 0xC0 /* Program Change */ =>
          programs += m.getChannel -> m.getData1
      }

      // Extract all messages in the current fragment, until the next fragment delimiter.
      val messages = midi.takeWhile {
        case (tick, msgs) => tick == start || msgs.forall {
          case m: MetaMessage if m.getType == 0x59     /* Key Change     */ => false
          case m: MetaMessage if m.getType == 0x58     /* Time Change    */ => false
          case m: MetaMessage if m.getType == 0x51     /* Tempo Change   */ => false
          case m: ShortMessage if m.getCommand == 0xC0 /* Program Change */ => false
          case _ => true
        }
      }

      messages.flatMap({
        case (tick, msgs) => msgs.collect {
          case e: ShortMessage if e.getCommand == 0x90 /* Note On */ =>
            (tick, e.getChannel, Note(Pitch(e.getData1 % 12, e.getData1 / 12), e.getData2 / 128.0))
          case e: ShortMessage if e.getCommand == 0x80 /* Note Off */ =>
            (tick, e.getChannel, Note(Pitch(e.getData1 % 12, e.getData1 / 12), 0.0))
        }
      }).groupBy(_._2).foreach({ case (channel, notes) =>
        val groups = new mutable.TreeMap[(Long, Long), Set[Note]]
        val active = mutable.Map.empty[Note, Long]

        notes.toSeq.sortBy(_._1).foreach { case (tick, _, note) =>
          // Deactivate the note if it is currently active.
          active.keys.find(_.pitch == note.pitch).foreach { last =>
            val duration = (active(last), tick)
            groups += duration -> (groups.getOrElse(duration, Set.empty) + last)
            active -= last
          }

          // Activate the note if and only if it has a non-zero volume.
          if (note.volume > 0) active += note -> tick
        }

        midi --= messages.keys

        // Construct a builder to generate chords from notes and midi tick durations.
        val voices = mutable.Buffer.empty[Voice]
        val ppb = 4 * sequence.getResolution

        while (groups.nonEmpty) {
          val chords = mutable.Buffer.empty[Chord]
          var prev = start
          var next = groups.from((prev, prev))

          // While we have not traversed the entire sequence of chords, iteratively find the next
          // chord whose start time is at least the previous time. Insert a rest before the next
          // chord if it does not immediately start after the previous chord. Then, insert the
          // chord itself and shift the previous time to the end of the next chord.
          while (next.nonEmpty) {
            val ((start, end), group) = next.head
            if (start > prev) chords += Chord(Set.empty, Duration((start - prev).toInt, ppb).reduce)
            chords += Chord(group, Duration((end - start).toInt, ppb).reduce)
            groups -= ((start, end))
            prev = end
            next = groups.from((prev, prev))
          }

          if (midi.nonEmpty) chords += Chord(Set.empty, Duration((midi.head._1 - prev).toInt, ppb).reduce)
          voices += Voice(chords)
        }

        // Construct a new phrase from the constructed voices and add it to the channel mapping.
        channels += channel -> Phrase(programs(channel), voices.toSet)
      })

      // Insert the fragment into the song.
      fragments += Fragment(key, tempo, channels.toMap)
    }

    Song(fragments)
  }

}